esi/src/lib.rs

```diff
@@ -23,210 +22,395 @@
-//!     let processor = Processor::new(config);
+//!     // If the response is HTML, we can parse it for ESI tags.
+//!     if beresp
+//!         .get_content_type()
+//!         .map(|c| c.subtype() == mime::HTML)
+//!         .unwrap_or(false)
+//!     {
+//!         let processor = esi::Processor::new(
+//!             // The original client request.
+//!             Some(req),
+//!             // Use the default ESI configuration.
+//!             esi::Configuration::default()
+//!         );
 //!
-//!     // Execute the ESI document using the client request as context
-//!     // and sending all requests to the backend `origin_1`.
-//!     processor.execute_esi(req, beresp, &|req| {
-//!         Ok(req.with_ttl(120).send("origin_1")?)
-//!     })?;
+//!         processor.process_response(
+//!             // The ESI source document. Note that the body will be consumed.
+//!             &mut beresp,
+//!             // Optionally provide a template for the client response.
+//!             Some(Response::from_status(StatusCode::OK).with_content_type(mime::TEXT_HTML)),
+//!             // Provide logic for sending fragment requests, otherwise the hostname
+//!             // of the request URL will be used as the backend name.
+//!             Some(&|req| {
+//!                 println!("Sending request {} {}", req.get_method(), req.get_path());
+//!                 Ok(Some(req.with_ttl(120).send_async("mock-s3")?))
+//!             }),
+//!             // Optionally provide a method to process fragment responses before they
+//!             // are streamed to the client.
+//!             Some(&|req, resp| {
+//!                 println!(
+//!                     "Received response for {} {}",
+//!                     req.get_method(),
+//!                     req.get_path()
+//!                 );
+//!                 Ok(resp)
+//!             }),
+//!         )?;
+//!     } else {
+//!         // Otherwise, we can just return the response.
+//!         beresp.send_to_client();
+//!     }
 //!
 //!     Ok(())
 //! }
 //! ```
 
 mod config;
+mod document;
 mod error;
 mod parse;
 
-use fastly::http::body::StreamingBody;
-use fastly::http::header;
-use fastly::{Body, Request, Response};
-use log::{debug, error, warn};
+use fastly::http::request::PendingRequest;
+use fastly::http::{header, Method, StatusCode, Url};
+use fastly::{mime, Body, Request, Response};
+use log::{debug, error};
 use quick_xml::{Reader, Writer};
-use std::io::Write;
+use std::collections::VecDeque;
+use std::io::{BufRead, Write};
 
+pub use crate::document::Element;
 use crate::error::Result;
 pub use crate::parse::{parse_tags, Event, Tag};
 
 pub use crate::config::Configuration;
 pub use crate::error::ExecutionError;
 
 /// An instance of the ESI processor with a given configuration.
-#[derive(Default)]
 pub struct Processor {
+    // The original client request metadata, if any.
+    original_request_metadata: Option<Request>,
```

Comment:
Can `original_request_metadata` be lumped into the `dispatch_fragment_request` callback? Maybe that callback should have type `FnMut(Url) -> Option<Request>`. Here's rationale for why I'm suggesting each of the changes:

- `FnMut` (instead of `Fn`) is the most general choice, allowing callers to capture mutable state in the closure. All `FnOnce` and `Fn` closures are also valid for use where `FnMut` is expected.
- Instead of feeding a caller-provided default `Request` into `build_fragment_request`, then feeding that into `dispatch_fragment_request`, you can let the callback build the request, using any state the caller has. The only data the library has that the callback can't already get some other way is the URL from the `include` tag attributes. I don't know whether that should be a `Url` or a `&str` though: should the library make the URL absolute or leave that to the callback?
- There are only two ways to get a `PendingRequest`, as far as I know: `send_async` and `send_async_streaming`. Assuming these fragment requests never need a large request body, you can simplify the interface by making the library call `send_async`, so the callback only has to construct a `Request`. Then the callback also doesn't need to report hostcall errors, so it doesn't need to return a `Result`. It only needs `Option` to indicate that a particular include should be dropped.

Reply:
I wanted to add a small usage convenience point: Since the only thing that the `original_request_metadata` parameter is used for is `clone_without_body`, it would be helpful to use a reference. As it is now (also in 0.3), taking ownership (needlessly) restricts what else the caller might want to do with the request object before/afterwards, and we have had to employ quite some contortions in our code at some points to make it work.

Reply:
Also, on the one hand, moving the construction of the request object to the callback sounds appealing, due to the added flexibility. On the other hand, it's not necessarily [all that straightforward](https://github.com/fastly/esi/issues/19) which state of the original request should be copied to the fragment request. So if this responsibility is left up to the caller, it would be helpful to at least document best practises here.